File: LogAndroid/TrustarApp/src/main/kotlin/jp/co/toukei/log/trustar/feature/nimachi/data/EditedTime.kt
package jp.co.toukei.log.trustar.feature.nimachi.data

class EditedTime(
        @JvmField val t: TimeItem,
        @JvmField val color: Int = 0x11FF0000
) : TimeItem(t.begin, t.end, t.type) {

    var justChanged = false
    var markDeleted = false
    var overrideBeginDate: DateItem? = null
    var overrideEndDate: DateItem? = null

    fun targetBeginDate(): DateItem? {
        return overrideBeginDate ?: begin
    }

    fun targetEndDate(): DateItem? {
        return overrideEndDate ?: end
    }

    override val beginDateString: String?
        get() = targetBeginDate()?.localDateString

    override val endDateString: String?
        get() = targetEndDate()?.localDateString

    fun hasChanged() = justChanged || markDeleted || overrideBeginDate != null || overrideEndDate != null
}

File: LogiOS/TrustarApp/Features/Nimachi/Data/EditedTime.swift
//
//  EditedTime.swift
//  TrustarApp
//
//  Created by DionSoftware on 2022/02/24.
//  Copyright © 2022 DionSoftware. All rights reserved.
//

class EditedTime: TimeItem {
    var t: TimeItem
    var color: Int
    
    var justChanged: Bool = false
    var markDeleted: Bool = false
    var overrideBeginDate: DateItem? = nil
    var overrideEndDate: DateItem? = nil
    
    //0x11ff0000
    init(t: TimeItem, color: Int = 0xffeeee) {
        self.t = t
        self.color = color
        super.init(begin: t.begin, end: t.end, type: t.type)
    }
    
    func targetBeginDate() -> DateItem? {
        return overrideBeginDate ?? begin
    }
    
    func targetEndDate() -> DateItem? {
        return overrideEndDate ?? end
    }
    
    override var beginDateString: String? {
        return targetBeginDate()?.localDateString
    }
    
    override var endDateString: String? {
        return targetEndDate()?.localDateString
    }
    
    func hasChanged() -> Bool {
        return justChanged || markDeleted ||
               overrideBeginDate != nil || overrideEndDate != nil
    }
}
